kernel-rs/src/fs/lfs/inode.rs

```diff
@@ -290,109 +319,101 @@ impl InodeGuard<'_, Lfs> {
-                let segment = tx.fs.segment();
-                if segment.is_full() {
-                    segment.commit(ctx);
-                }
-            }
-
-            let mut bp = hal().disk().read(self.dev, indirect, ctx);
-            let (prefix, data, _) = unsafe { bp.deref_inner_mut().data.align_to_mut::<u32>() };
+            // Read the indirect block.
+            let indirect = self.deref_inner().addr_indirect;
+            assert!(indirect != 0, "bmap: out of range");
+            let bp = hal().disk().read(self.dev, indirect, ctx);
+            // Get the address.
+            let (prefix, data, _) = unsafe { bp.deref_inner().data.align_to::<u32>() };
             debug_assert_eq!(prefix.len(), 0, "bmap: Buf data unaligned");
             let addr = data[bn];
-            let buf = if write {
-                let tx = tx_opt.expect("bmap: out of range");
-                let (buf, addr) = if addr == 0 {
-                    // Allocate an empty block.
-                    tx.balloc(self.inum, bn as u32, ctx)
-                } else {
-                    let (mut buf, new_addr) = tx
-                        .fs
-                        .segment()
-                        .get_or_add_data_block(self.inum, bn as u32, ctx)
-                        .unwrap();
-                    if new_addr != addr {
-                        // Copy from old block to new block.
-                        let old_buf = hal().disk().read(self.dev, addr, ctx);
-                        // SAFETY: The old data block's content will not be used from now on.
-                        unsafe {
-                            core::ptr::copy(
-                                &raw const old_buf.deref_inner().data,
-                                &raw mut buf.deref_inner_mut().data,
-                                1,
-                            );
-                        }
-                        old_buf.free(ctx);
-                    }
-                    (buf, new_addr)
-                };
-                data[bn] = addr;
-                buf
-            } else {
-                assert!(addr != 0, "bmap: out of range");
-                hal().disk().read(self.dev, addr, ctx)
-            };
             bp.free(ctx);
-            buf
+            // Read the indirect data block.
+            assert!(addr != 0, "bmap: out of range");
+            hal().disk().read(self.dev, addr, ctx)
+        }
+    }
+
+    /// Returns the `bn`th data block of the inode and its (possibly new) disk block number.
+    /// The given `addr` is the (possibly old) disk block number of the `bn`th data block.
+    /// * If `addr == 0`, allocates an empty disk block and returns it.
+    /// * If `addr != 0`, the content of the data at `addr` is copied to the returned block if a new block was allocated.
+    ///
+    /// # Note
+    ///
+    /// You should make sure the segment has an empty block before calling this.
```

Comment:
Will it be better to check this inside this function?

Reply:
Since `InodeGuard::{writable_data_block_inner, writable_indirect_block}` is only used inside the `InodeGuard::writable_data_block`, and since those two methods must be called always after checking that we have at least two blocks on the segment, I think its better to just panic instead of runtime check inside those functions.